cleverhans/attacks.py

```diff
@@ -1,4 +1,351 @@
+from abc import ABCMeta
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, pred, back='tf', sess=None, params={}):
+        """
+        :param pred: The model's symbolic output.
+        :param back: The backend to use. Either 'tf' (default) or 'th'.
+        :param params: Parameter dictionary used by child classes.
+        """
+        if not back == 'tf' or back == 'th':
+            raise Exception("Backend argument must either be 'tf' or 'th'.")
+        if back == 'tf' and sess is None:
+            raise Exception("A tf session was not provided in sess argument.")
+        if back == 'th' and sess is not None:
+            raise Exception("A session should not be provided when using th.")
+
+        # Prepare attributes
+        self.x = None
+        self.pred = pred
+        self.back = back
+        self.sess = sess
+        self.nb_calls_generate = 0
+        self.default_graph = None
+
+    def generate(self, x):
+        """
+        Generate the attack's symbolic graph for adversarial examples. This
+        method should be overriden in any child class that implements an
+        attack that is expressable symbolically.
+        :param x: The model's symbolic inputs.
+        :return: A symbolic representation of the adversarial examples.
+        """
+        # Keep track of the number of calls to warn when more than one default
+        # graph to run if generating adversarial examples numerically
+        self.nb_calls_generate += 1
+
+        self.x = x
+
+        if self.back == 'tf':
+            import tensorflow as tf
+            wrapper = tf.py_func(self.generate_np, [self.x], tf.float32)
+            if self.nb_calls_generate == 1:
+                self.default_graph = wrapper
+            return wrapper
+        else:
+            # TODO()
+            return False
+
+    def generate_np(self, X, params={'batch_size': 128}):
+        """
+        Generate adversarial examples and return them as a Numpy array. This
+        method should be overriden in any child class that implements an attack
+        that is not fully expressed symbolically.
+        :param X: A Numpy array with the original inputs.
+        :param params: Parameter dictionary used by child classes.
+        :return: A Numpy array holding the adversarial examples.
+        """
+        if self.default_graph is None:
+            error_string = "The attack symbolic graph was not generated."
+            raise NotImplementedError(error_string)
+        if self.nb_calls_generate > 1:
+            warnings.warn("Attack was generated symbolically multiple "
```

Comment:
The user shouldn't really know or care whether `generate_np` uses `generate` internally.
`generate_np` should just work.

Internally, it should make its own graph once and then re-use that every time `generate_np` is called. Something like

    if not hasattr(self, "_x"):
      self._x = <make placeholder>
      self._x_adv = self.generate(self._x, <other args>)
    return self._x_adv.eval(feed_dict={self._x: x})